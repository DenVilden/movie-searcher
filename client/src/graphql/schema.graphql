extend type Query {
  favorites: [String!]!
}

extend type MovieInfo {
  isInFavorites: Boolean!
}

extend type Mutation {
  addOrRemoveFromFavorites(id: String!): [String!]!
}
