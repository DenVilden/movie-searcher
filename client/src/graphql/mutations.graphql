mutation AddOrRemoveFromFavorites($id: String!) {
  addOrRemoveFromFavorites(id: $id) @client
}
