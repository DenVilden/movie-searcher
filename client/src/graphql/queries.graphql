query GetMovieInfo($id: String!) {
  movieInfo(id: $id) {
    isInFavorites @client
    id
    backdrop_path
    poster_path
    title
    overview
    budget
    revenue
    vote_average
    release_date
    similar {
      results {
        id
        title
        release_date
        poster_path
      }
    }
  }
}

query GetMovies {
  upcoming {
    total_pages
    page
    results {
      id
      title
      release_date
      poster_path
    }
  }
  topRated {
    total_pages
    page
    results {
      id
      title
      vote_average
      poster_path
    }
  }
}

query GetUpcoming($page: Int) {
  upcoming(page: $page) {
    total_pages
    page
    results {
      id
      title
      release_date
      poster_path
    }
  }
}

query GetTopRated($page: Int) {
  topRated(page: $page) {
    total_pages
    page
    results {
      id
      title
      vote_average
      poster_path
    }
  }
}

query GetMoviesSearch($query: String!, $cursor: Int) {
  moviesSearch(query: $query, cursor: $cursor) {
    cursor
    hasMore
    results {
      id
      title
    }
  }
}

query GetFavorites {
  favorites @client
}
