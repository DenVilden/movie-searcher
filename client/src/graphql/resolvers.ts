import { ApolloClient, NormalizedCacheObject } from "@apollo/client";
import {
  GetFavoritesQuery,
  GetFavoritesDocument,
  Resolvers,
} from "./__generated__";

const resolvers: Resolvers<ApolloClient<NormalizedCacheObject>> = {
  MovieInfo: {
    isInFavorites: (movie, __, { cache }) => {
      const queryResult = cache.readQuery<GetFavoritesQuery>({
        query: GetFavoritesDocument,
      });

      return queryResult!.favorites.includes(movie.id.toString());
    },
  },
  Mutation: {
    addOrRemoveFromFavorites: (_, { id }, { cache }) => {
      const queryResult = cache.readQuery<GetFavoritesQuery>({
        query: GetFavoritesDocument,
      });

      const { favorites } = queryResult!;

      const newFavorites = favorites.includes(id)
        ? favorites.filter((favId) => favId !== id)
        : [...favorites, id];

      cache.writeQuery({
        query: GetFavoritesDocument,
        data: { favorites: newFavorites },
      });

      return newFavorites;
    },
  },
};

export default resolvers;
