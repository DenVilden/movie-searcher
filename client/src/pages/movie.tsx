import React, { useEffect } from 'react';
import { Slide, LinearProgress } from '@material-ui/core';
import { useRouter } from 'next/router';
import MovieInfo from '../components/MovieInfo/MovieInfo';
import MoviesBox from '../components/MoviesBox/MoviesBox';
import ErrorMessage from '../containers/ErrorMessage';
import {
  GetMovieInfoDocument,
  useGetMovieInfoLazyQuery,
} from '../generated/queries.generated';
import {
  useSetInputValueMutation,
  useAddOrRemoveFromFavoritesMutation,
} from '../generated/mutations.generated';
import { withApollo } from '../lib/withApollo';
import Header from '../containers/Header';

export const MoviePage = () => {
  const {
    query: { id },
  } = useRouter();

  const [fetchMovies, { loading, error, data }] = useGetMovieInfoLazyQuery();

  const [setInputValue] = useSetInputValueMutation();

  useEffect(() => {
    if (id) {
      fetchMovies({
        variables: { id: id as string },
      });
    }
    setInputValue({ variables: { value: '' } });
  }, [id, fetchMovies, setInputValue]);

  const [
    addOrRemoveFromFavorites,
    favoritesMutationOptions,
  ] = useAddOrRemoveFromFavoritesMutation();

  if (error) return <ErrorMessage error={error} />;

  return (
    <>
      <Header />
      {loading || !data?.movieInfo ? (
        <LinearProgress color="secondary" />
      ) : (
        <Slide direction="up" in>
          <div>
            <MovieInfo
              addOrRemoveFromFavorites={() =>
                addOrRemoveFromFavorites({
                  variables: { id: id as string },
                  refetchQueries: [
                    {
                      query: GetMovieInfoDocument,
                      variables: { id: id as string },
                    },
                  ],
                })
              }
              loading={favoritesMutationOptions.loading}
              movie={data.movieInfo}
            />
            {!!data.movieInfo.similar.results.length && (
              <MoviesBox
                movies={data.movieInfo.similar.results}
                title="Similar Movies"
              />
            )}
          </div>
        </Slide>
      )}
    </>
  );
};

export default withApollo()(MoviePage);
