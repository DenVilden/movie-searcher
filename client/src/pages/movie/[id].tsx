import React, { useEffect } from 'react';
import { Slide, LinearProgress } from '@material-ui/core';
import { useRouter } from 'next/router';
import MovieInfo from '../../components/MovieInfo/MovieInfo';
import MoviesBox from '../../components/MoviesBox/MoviesBox';
import ErrorMessage from '../../containers/ErrorMessage';
import {
  useGetMovieInfoQuery,
  useSetInputValueMutation,
  useAddOrRemoveFromFavoritesMutation,
  GetMovieInfoDocument,
} from '../../__generated__';

const MoviePage = () => {
  const {
    query: { id },
  } = useRouter();

  const { loading, error, data } = useGetMovieInfoQuery({
    variables: { id: id as string },
  });

  const [setInputValue] = useSetInputValueMutation();

  const [addOrRemoveFromFavorites] = useAddOrRemoveFromFavoritesMutation();

  useEffect(() => {
    setInputValue({ variables: { value: '' } });
  }, [setInputValue, id]);

  if (loading) return <LinearProgress color="secondary" />;

  if (error) return <ErrorMessage error={error} />;

  if (!data?.movieInfo) throw new Error('Not found');

  return (
    <Slide direction="up" in>
      <div>
        <MovieInfo
          addOrRemoveFromFavorites={() =>
            addOrRemoveFromFavorites({
              variables: { id: id as string },
              refetchQueries: [
                {
                  query: GetMovieInfoDocument,
                  variables: { id: id as string },
                },
              ],
            })
          }
          movie={data.movieInfo}
        />
        {!!data.movieInfo.similar.results.length && (
          <MoviesBox
            movies={data.movieInfo.similar.results}
            title="Similar Movies"
          />
        )}
      </div>
    </Slide>
  );
};

export default MoviePage;
