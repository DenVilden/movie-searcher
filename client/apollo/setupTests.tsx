import { render } from '@testing-library/react';
import { MockedProvider, MockedResponse } from '@apollo/client/testing';
import { ThemeProvider } from 'styled-components';
import { theme } from '../pages/_app';

// eslint-disable-next-line jest/prefer-spy-on
HTMLElement.prototype.scrollIntoView = jest.fn();

type RenderApolloOptions = {
  mocks?: MockedResponse[];
  addTypeName?: boolean;
};

export const renderApollo = (
  element: React.ReactElement,
  { mocks = [], addTypeName = false }: RenderApolloOptions = {},
) =>
  render(
    <MockedProvider addTypename={addTypeName} mocks={mocks}>
      <ThemeProvider theme={theme}>{element}</ThemeProvider>
    </MockedProvider>,
  );

export * from '@testing-library/react';
