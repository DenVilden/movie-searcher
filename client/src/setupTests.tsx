import React from 'react';
import { render } from '@testing-library/react';
import { MockedProvider, MockedResponse } from '@apollo/client/testing';
import { ThemeProvider } from 'styled-components';
import { theme } from './pages/_app';
import defaultResolvers from './graphql/resolvers';

window.HTMLElement.prototype.scrollIntoView = jest.fn();

type RenderApolloOptions = {
  mocks?: MockedResponse[];
  addTypeName?: boolean;
  cache?: any;
  resolvers?: any;
};

export const renderApollo = (
  element: React.ReactElement,
  { mocks, addTypeName, resolvers, cache }: RenderApolloOptions = {}
) =>
  render(
    <MockedProvider
      addTypename={addTypeName}
      cache={cache}
      mocks={mocks}
      resolvers={resolvers || defaultResolvers}
    >
      <ThemeProvider theme={theme}>{element}</ThemeProvider>
    </MockedProvider>
  );

export * from '@testing-library/react';
