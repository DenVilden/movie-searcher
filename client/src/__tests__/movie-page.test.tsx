import { MoviePage } from "../pages/movie/[id]";
import { GetMovieInfoDocument } from "../graphql";
import { renderApollo, fireEvent } from "../setupTests";

jest.mock("next/router", () => ({
  useRouter: () => ({
    query: {
      id: "1",
    },
  }),
}));

const mocks = [
  {
    request: {
      query: GetMovieInfoDocument,
      variables: {
        id: "1",
      },
    },
    result: {
      data: {
        movieInfo: {
          id: 1,
          backdrop_path: null,
          poster_path: null,
          title: "test",
          overview: "test data",
          budget: "0",
          revenue: "0",
          vote_average: 5,
          release_date: "2020",
          similar: {
            results: [
              {
                id: 1,
                title: "test",
                release_date: "2020",
                poster_path: null,
              },
            ],
          },
        },
      },
    },
  },
];

describe("moviePage", () => {
  it("should render error state", async () => {
    const mockError = [
      {
        request: {
          query: GetMovieInfoDocument,
          variables: {
            id: "1",
          },
        },
        error: new Error("an error has occurred"),
      },
    ];

    const { findByText } = renderApollo(<MoviePage />, {
      mocks: mockError,
    });

    const errorElement = await findByText(/an error has occurred/i);

    expect(errorElement).toBeTruthy();
  });

  it("should toggle favorites", async () => {
    const mock = [mocks[0], mocks[0], mocks[0]];

    const { findByTestId, findByText } = renderApollo(<MoviePage />, {
      mocks: mock,
    });

    const favoritesButton = await findByTestId("favorites-button");

    fireEvent.click(favoritesButton);

    const removeButton = await findByText("Remove from favorites");

    expect(removeButton).toBeTruthy();

    fireEvent.click(favoritesButton);

    const addButton = await findByText("Add to favorites");

    expect(addButton).toBeTruthy();
  });
});
