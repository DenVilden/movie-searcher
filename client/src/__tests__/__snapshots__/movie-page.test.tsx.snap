// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`moviePage should take a snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiLinearProgress-root MuiLinearProgress-colorSecondary MuiLinearProgress-indeterminate"
    role="progressbar"
  >
    <div
      class="MuiLinearProgress-bar MuiLinearProgress-barColorSecondary MuiLinearProgress-bar1Indeterminate"
    />
    <div
      class="MuiLinearProgress-bar MuiLinearProgress-bar2Indeterminate MuiLinearProgress-barColorSecondary"
    />
  </div>
</DocumentFragment>
`;
